### DISCUSSION TITLE
Adding cupy support

### DISCUSSION THREAD

**Discussion by JoOkuma** [wjakob/nanobind#510] [Ideas]
Hello, first of all, thanks for the amazing package.

Any thoughts on adding cupy support?
I think it should be straightforward since they already [support DLpack](https://docs.cupy.dev/en/stable/reference/generated/cupy.ndarray.html#cupy.ndarray.toDlpack).

I can implement it, given some pointers.

If it's already supported, I might need some help 😅.
I tried passing a cupy.ndarray and DLpack Capsule using ndarray with `nb::device::cuda` and it didn't work.

**Comment by wjakob** [comment]
Does `cupy.ndarray` have an `__dlpack__` member? If so, I'm surprised that it does not work. Otherwise, what's the way of extracting the DLPack capsule?

**Reply by JoOkuma** [reply]
Yes, [they do](https://github.com/cupy/cupy/blob/027d11cef08efb7e2fe50d3db6591aee5ce021ad/cupy/array_api/_array_object.py#L457).

My impression is that nanobind type checks and fails before reaching the `__dlpack__` call.
```python
E           TypeError: _cu_func(): incompatible function arguments. The following argument types are supported:
E               1. _cu_func(arg0: ndarray[dtype=uint32, shape=(*, *), order='C', device='cuda'], /) -> numpy.ndarray[dtype=uint64]
E    
E           Invoked with types: cupy.ndarray
```

**Reply by JoOkuma** [reply]
Hi @wjakob, any additional thoughts on this?

**Reply by wjakob** [reply]
I don't use CuPy myself but am open to changes. Would you be able to create a PR that extends the framework detection to handle this case?

**Reply by JoOkuma** [reply]
Sure, I created #594. Let me know if I missed something, thanks!
